classes/kohana/encrypt.php

```diff
@@ -1,213 +1,121 @@
 	}
 
 	/**
-	 * Encrypts a string and returns an encrypted string that can be decoded.
-	 *
-	 *     $data = $encrypt->encode($data);
+	 * Decrypts the encoded base 64 data to it's original value.
+	 * Returns false if the decryption fails.
 	 *
-	 * The encrypted binary data is encoded using [base64](http://php.net/base64_encode)
-	 * to convert it to a string. This string can be stored in a database,
-	 * displayed, and passed using most other means without corruption.
+	 * @param string $data
 	 *
-	 * @param   string  $data   data to be encrypted
-	 * @return  string
+	 * @return string|false
 	 */
-	public function encode($data)
+	public function decode($data)
 	{
-		// Set the rand type if it has not already been set
-		if (Encrypt::$_rand === NULL)
-		{
-			if (Kohana::$is_windows)
-			{
-				// Windows only supports the system random number generator
-				Encrypt::$_rand = MCRYPT_RAND;
-			}
-			else
-			{
-				if (defined('MCRYPT_DEV_URANDOM'))
-				{
-					// Use /dev/urandom
-					Encrypt::$_rand = MCRYPT_DEV_URANDOM;
-				}
-				elseif (defined('MCRYPT_DEV_RANDOM'))
-				{
-					// Use /dev/random
-					Encrypt::$_rand = MCRYPT_DEV_RANDOM;
-				}
-				else
-				{
-					// Use the system random number generator
-					Encrypt::$_rand = MCRYPT_RAND;
-				}
-			}
-		}
+		$data = base64_decode($data, true);
 
-		if (Encrypt::$_rand === MCRYPT_RAND)
+		if (!$data)
 		{
-			// The system random number generator must always be seeded each
-			// time it is used, or it will not produce true random results
-			mt_srand();
+			return false;
 		}
 
-		// Create a random initialization vector of the proper size for the current cipher
-		$iv = mcrypt_create_iv($this->_iv_size, Encrypt::$_rand);
+		// Nonce is twice the size when converted to hexadecimal.
+		$nonceSize = SODIUM_CRYPTO_SECRETBOX_NONCEBYTES * 2;
+
+		// Extract the nonce (hexadecimal) from the data (hexadecimal).
+		$nonce = substr($data, 0, $nonceSize);
+
+		// Remove the nonce from the data.
+		$data = substr($data, $nonceSize);
 
-		// Encrypt the data using the configured options and generated iv
-		$data = mcrypt_encrypt($this->_cipher, $this->_key, $data, $this->_mode, $iv);
+		$encryptedData = sodium_crypto_secretbox_open(hex2bin($data), hex2bin($nonce), $this->key);
+
+		if (!$encryptedData)
+		{
+			return false;
+		}
 
-		// Use base64 encoding to convert to a string
-		return base64_encode($iv.$data);
+		return $encryptedData;
 	}
 
 	/**
-	 * Decrypts an encoded string back to its original value.
+	 * Returns a singleton instance of Encrypt. An encryption key must be
+	 * provided in your "encrypt" configuration file.
 	 *
-	 *     $data = $encrypt->decode($data);
+	 * @param string $name configuration group name
 	 *
-	 * @param   string  $data   encoded string to be decrypted
-	 * @return  FALSE   if decryption fails
-	 * @return  string
+	 * @return self
 	 */
-	public function decode($data)
+	public static function instance($name = null)
```

Comment:
Can we put the methods back in the same order so this PR is easier to review please.